### PR TITLE
Add is_public to Category

### PR DESCRIPTION
**Pull request by jaroel**
Allows for disabling categories without deleting them.
This is so shop owners can build up new categories without them being available for the public.

### CODE REVIEW

**Comment by solarissmoke** [comment]
This looks good to me - want to hold off merging for a few days until we ship v2.0.

**Comment by jaroel** [comment]
Not a problem, thanks!

File: src/oscar/apps/catalogue/abstract_models.py [hunk 1]
```diff
@@ -77,6 +77,16 @@ def has_attributes(self):
         return self.attributes.exists()
 
 
+class CategoryQuerySet(MP_NodeQuerySet):
+    def public(self):
```
Comment (solarissmoke) [inline src/oscar/apps/catalogue/abstract_models.py]: Can we change this to `browsable()` to be consistent with `ProductQuerySet`, and add a docstring to this method?

Comment (jaroel) [reply src/oscar/apps/catalogue/abstract_models.py]: Thanks for the feedback. I've renamed `public` to `browsable`.

**Comment by specialunderwear** [comment]
#3160 is a problem with this code

File: src/oscar/apps/catalogue/abstract_models.py [hunk 2]
```diff
@@ -77,6 +77,19 @@ def has_attributes(self):
         return self.attributes.exists()
 
 
+class CategoryQuerySet(MP_NodeQuerySet):
+    def browsable(self):
+        """
+        Excludes non-public categories
+        """
+        return self.filter(is_public=True)
```
Comment (specialunderwear) [inline src/oscar/apps/catalogue/abstract_models.py]: This is not sufficient, all subcategories of non-public categories must be filtered out as well. If not bad thing happen, see #3160

Comment (specialunderwear) [reply src/oscar/apps/catalogue/abstract_models.py]: fixed in 3de975c

**Comment by specialunderwear** [comment]
This is now good to merge, just needs approval from @solarissmoke

**Comment by specialunderwear** [comment]
This PR needs work because it is too slow.

**Comment by specialunderwear** [comment]
@jaroel can you update this PR with your performance fixes on postgres?

**Comment by specialunderwear** [comment]
These changes do indeed solve the performance problem that was only apparent on postgresql.

File: src/oscar/apps/catalogue/receivers.py [hunk 3]
```diff
@@ -1,16 +1,19 @@
...
+
+from oscar.core.loading import get_model
+
+Category = get_model("catalogue", "Category")
 
 if settings.OSCAR_DELETE_IMAGE_FILES:
     from django.db import models
-    from django.db.models.signals import post_delete
 
     from oscar.core.loading import get_model
```
Comment (solarissmoke) [inline src/oscar/apps/catalogue/receivers.py]: Duplicate import.

File: src/oscar/apps/catalogue/abstract_models.py [hunk 4]
```diff
@@ -34,6 +36,41 @@
...
+     Would be about the same as the python statement::
+
+        henk.startswith("koe")
+
+    ReverseStartsWith will flip the right and left hand side of the expression,
+    effectively making this the same query as::
+
+    "koe".startswith(henk)
+
+    This is used by the range query below, where we need to flip select children
```
Comment (solarissmoke) [inline src/oscar/apps/catalogue/abstract_models.py]: This comment is no longer valid because the class has moved. Let's remove it.

**Comment by jaroel** [comment]
Agree - I'll get it sorted asap

**Comment by jaroel** [comment]
Awesome, thanks!

**Comment by solarissmoke** [comment]
I've updated and rebased this PR. Note that I've changed how the ``Category.objects`` manager is defined to be consistent with what we do for `Product` - we don't need a separate manager class. I've also moved the queryset into `managers.py` so that it can be overridden.

File: src/oscar/apps/catalogue/abstract_models.py [hunk 5]
```diff
@@ -158,9 +207,34 @@ def save(self, *args, **kwargs):
...
         super().save(*args, **kwargs)
 
+    def set_ancestors_are_public(self):
+        # Update ancestors_are_public for the sub tree.
+        # note: This doesn't trigger a new save for each instance, rather
+        # just a SQL update.
+        included_in_non_public_subtree = self.__class__.objects.filter(
+            is_public=False, path__rstartswith=OuterRef("path"), depth__lt=OuterRef("depth")
+        )
+        self.get_descendants_and_self().update(
```
Comment (iamsobanjaved) [inline src/oscar/apps/catalogue/abstract_models.py:219]: @jaroel is it possible to change this query which is mysql compatible as well.

> MySQL doesn't allow updating the table you are already using in an inner select as the update criteria.

Comment (jaroel) [reply src/oscar/apps/catalogue/abstract_models.py:219]: Dunno, I don't use mysql. What did you already try but didn't work?

**Pull request merged**